examples/android/clientcache/app/src/main/java/io/grpc/clientcacheexample/SafeMethodCachingInterceptor.java

```diff
@@ -0,0 +1,297 @@
+            }
+          };
+
+      @Override
+      public void put(Key key, Value value) {
+        lruCache.put(key, value);
+      }
+
+      @Override
+      public Value get(Key key) {
+        return lruCache.get(key);
+      }
+
+      @Override
+      public void remove(Key key) {
+        lruCache.remove(key);
+      }
+
+      @Override
+      public void clear() {
+        lruCache.evictAll();
+      }
+    };
+  }
+
+  public static SafeMethodCachingInterceptor newSafeMethodCachingInterceptor(Cache cache) {
+    return newSafeMethodCachingInterceptor(cache, DEFAULT_MAX_AGE_SECONDS);
+  }
+
+  public static SafeMethodCachingInterceptor newSafeMethodCachingInterceptor(
+      Cache cache, int defaultMaxAge) {
+    return new SafeMethodCachingInterceptor(cache, defaultMaxAge);
+  }
+
+  private static int DEFAULT_MAX_AGE_SECONDS = 3600;
+
+  private static final Metadata.Key<String> CACHE_CONTROL_KEY =
+      Metadata.Key.of("cache-control", Metadata.ASCII_STRING_MARSHALLER);
+
+  private static final Splitter CACHE_CONTROL_SPLITTER =
+      Splitter.on(',').trimResults().omitEmptyStrings();
+
+  private final Cache internalCache;
+  private final int defaultMaxAge;
+
+  private SafeMethodCachingInterceptor(Cache cache, int defaultMaxAge) {
+    this.internalCache = cache;
+    this.defaultMaxAge = defaultMaxAge;
+  }
+
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+      final MethodDescriptor<ReqT, RespT> method, final CallOptions callOptions, Channel next) {
+    // Currently only unary methods can be marked safe, but check anyways.
+    if (!method.isSafe() || method.getType() != MethodDescriptor.MethodType.UNARY) {
+      return next.newCall(method, callOptions);
+    }
+
+    final String fullMethodName = method.getFullMethodName();
+
+    return new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(
+        next.newCall(method, callOptions)) {
+      private Listener<RespT> interceptedListener;
+      private Key requestKey;
+      private boolean cacheResponse = true;
+      private boolean cacheOptionsCausedCancellation;
+
+      @Override
+      public void start(Listener<RespT> responseListener, Metadata headers) {
+        interceptedListener =
+            new ForwardingClientCallListener.SimpleForwardingClientCallListener<RespT>(
+                responseListener) {
+              private Deadline deadline;
+              private int maxAge = -1;
+
+              @Override
+              public void onHeaders(Metadata headers) {
+                Iterable<String> cacheControlHeaders = headers.getAll(CACHE_CONTROL_KEY);
+                if (cacheResponse && cacheControlHeaders != null) {
+                  for (String cacheControlHeader : cacheControlHeaders) {
+                    for (String directive : CACHE_CONTROL_SPLITTER.split(cacheControlHeader)) {
+                      if (directive.equalsIgnoreCase("no-cache")) {
+                        cacheResponse = false;
```

Comment:
this is scary.   I don't know if I want people copy pasting this .  Doesn't OkHTTP have one you can re appropriate?

Reply:
From our offline conversation with @ejona86 my takeaway was that this limited cache-control parsing is OK for inclusion in an example app (perhaps with a more prominent indication of what this does and does not handle in the documentation). There are obstacles to re-appropriating the (internal OkHttp API) cache used by OkHttp, so for this basic example I'd prefer to stick with the current manual parsing.